pyccel/ast/operators.py

```diff
@@ -59,25 +59,35 @@
 def broadcast(shape_1, shape_2):
     """ This function broadcast two shapes using numpy broadcasting rules """
 
+    from pyccel.ast.sympy_helper import pyccel_to_sympy
+
     a = len(shape_1)
     b = len(shape_2)
     if a>b:
-        new_shape_2 = (1,)*(a-b) + tuple(shape_2)
+        new_shape_2 = (LiteralInteger(1),)*(a-b) + tuple(shape_2)
         new_shape_1 = shape_1
     elif b>a:
-        new_shape_1 = (1,)*(b-a) + tuple(shape_1)
+        new_shape_1 = (LiteralInteger(1),)*(b-a) + tuple(shape_1)
         new_shape_2 = shape_2
     else:
         new_shape_2 = shape_2
         new_shape_1 = shape_1
 
     new_shape = []
     for e1,e2 in zip(new_shape_1, new_shape_2):
-        if e1 == e2:
+        sy_e1 = pyccel_to_sympy(e1, {}, set())
+        sy_e2 = pyccel_to_sympy(e2, {}, set())
```

Comment:
I think it would be a good idea to pass the same `symbol_map` and `used_names` set to these 2 calls. That way we should be able to raise an error if we do:
```
def f(k):
    a = np.ones(k)
    b = np.ones(np.shape(a)[0]+2)
    c = a + b
```
Currently we get:
`sy_e1 = tmp_size_0001`
and
`sy_e2 = tmp_size_0001`
As the operator is not conserved we cannot raise an error currently, but this example still demonstrates the danger of not using the same dictionary and set

Reply:
Once this is implemented it would probably be worth modifying [these lines](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/variable.py#L220-L225) to accept any `PyccelAstNode`. The simplification of the shape is probably no longer necessary (so we could get rid of `alloc_shape`), although the `Shape` function should also be modified to always return a PyccelArrayShape rather than the element of the shape. All of these corrections are probably beyond the scope of this PR. But if you make minor changes locally (e.g. add `Variable` and `PyccelOperator` to the tuple at [this line](https://github.com/pyccel/pyccel/blob/master/pyccel/ast/variable.py#L220) ) then you can test your implementation in a more strenuous setting. E.g. for the example above I get:
`sy_e1 = k`
`sy_e2 = k + 2`
which your current implementation considers is acceptable

Reply:
I just realised there's a much simpler example that doesn't work currently and should already:
```
import numpy as np

def f(k : int):
    a = np.ones(k)
    c = a[1:] + a[2:]
```

Reply:
This sympy page may be helpful for working out if you have a system of equations with no solution: https://docs.sympy.org/latest/modules/solvers/solvers.html